subprojects/language-native/src/main/java/org/gradle/language/nativeplatform/tasks/DiscoverInputs.java

```diff
@@ -0,0 +1,203 @@
+import org.gradle.api.internal.file.collections.DirectoryFileTreeFactory;
+import org.gradle.api.provider.PropertyState;
+import org.gradle.api.tasks.Input;
+import org.gradle.api.tasks.InputFiles;
+import org.gradle.api.tasks.OutputFile;
+import org.gradle.api.tasks.SkipWhenEmpty;
+import org.gradle.api.tasks.TaskAction;
+import org.gradle.api.tasks.incremental.IncrementalTaskInputs;
+import org.gradle.cache.PersistentStateCache;
+import org.gradle.internal.hash.FileHasher;
+import org.gradle.language.nativeplatform.internal.incremental.CompilationState;
+import org.gradle.language.nativeplatform.internal.incremental.CompilationStateCacheFactory;
+import org.gradle.language.nativeplatform.internal.incremental.DefaultSourceIncludesParser;
+import org.gradle.language.nativeplatform.internal.incremental.DefaultSourceIncludesResolver;
+import org.gradle.language.nativeplatform.internal.incremental.IncrementalCompilation;
+import org.gradle.language.nativeplatform.internal.incremental.IncrementalCompileFilesFactory;
+import org.gradle.language.nativeplatform.internal.incremental.IncrementalCompileProcessor;
+import org.gradle.language.nativeplatform.internal.incremental.sourceparser.CSourceParser;
+import org.gradle.language.nativeplatform.internal.incremental.sourceparser.RegexBackedCSourceParser;
+import org.gradle.util.CollectionUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Inject;
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.OutputStreamWriter;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Responsible for discovering headers which are inputs to a native compilation task.
+ *
+ * @since 4.3
+ */
+@NonNullApi
+@Incubating
+public class DiscoverInputs extends DefaultTask {
+    private final Logger logger = LoggerFactory.getLogger(DiscoverInputs.class);
+
+    private final ConfigurableFileCollection includes;
+    private final ConfigurableFileCollection source;
+    private ImmutableList<String> includePaths;
+    private PropertyState<Boolean> importsAreIncludes;
+    private final RegularFileVar discoveredInputs;
+
+    private CSourceParser sourceParser;
+    private final FileHasher hasher;
+    private final CompilationStateCacheFactory compilationStateCacheFactory;
+    private final DirectoryFileTreeFactory directoryFileTreeFactory;
+
+    @Inject
+    public DiscoverInputs(FileHasher hasher, CompilationStateCacheFactory compilationStateCacheFactory, DirectoryFileTreeFactory directoryFileTreeFactory) {
+        this.hasher = hasher;
+        this.compilationStateCacheFactory = compilationStateCacheFactory;
+        this.directoryFileTreeFactory = directoryFileTreeFactory;
+        this.includes = getProject().files();
+        this.source = getProject().files();
+        this.sourceParser = new RegexBackedCSourceParser();
```

Comment:
The discovery in its current state works seems to work for all native compilation tasks which use discovered inputs, so it would also work for `ObjectiveC`. This is why I didn't give the task a language specific name.

Reply:
Does it handle `#import` as well as `#include`? If that's the case, we should not call it `CSourceParser`. Maybe something like `CStyleSourceParser`? It's also not a generic "source" parser, but something that detects includes, so maybe `CStyleIncludeDecetor`?

Reply:
The type has been present before. I would not rename it right now.

Reply:
:+1: